src/adera/Machines/RCSController.cpp

```diff
@@ -134,62 +110,104 @@ void SysMachineRCSController::update_construct(ActiveScene &rScene)
+                auto const *rbAncestor
+                        = SysPhysics_t::try_get_or_find_rigidbody_ancestor(
+                            rScene, ent);
+                auto const &compRb = rScene.reg_get<ACompRigidBody_t>(
+                            rbAncestor->m_ancestor);
+                auto const &compTf = rScene.reg_get<ACompTransform>(
+                            rbAncestor->m_ancestor);
+
+                Matrix4 transform = rbAncestor->m_relTransform;
+
+                // TODO: RCS translation is not currently implemented, only
+                // rotation
+                Vector3 commandTransl = Vector3{0.0f};
+                Vector3 commandRot = nodeCommand.m_state.m_value.m_attitude;
+                Vector3 thrusterPos = transform.translation()
+                                    - compRb.m_centerOfMassOffset;
+                Vector3 thrusterDir = transform.rotation()
+                                    * Vector3{0.0f, 0.0f, 1.0f};
+
+                if (commandRot.length() > 0.0f || commandTransl.length() > 0.0f)
+                {
+                    influence = thruster_influence(thrusterPos, thrusterDir,
+                                                   commandTransl, commandRot);
+                }
+            }
+        }
+
+        auto const *panelPercent = rScene.get_registry()
+                .try_get< ACompWirePanel<Percent> >(ent);
 
-            // TODO: RCS translation is not currently implemented, only rotation
-            Vector3 commandTransl = Vector3{0.0f};
-            Vector3 commandRot = attCtrl->m_attitude;
-            Vector3 thrusterPos = transform.translation() - rCompRb.m_centerOfMassOffset;
-            Vector3 thrusterDir = transform.rotation() * Vector3{0.0f, 0.0f, 1.0f};
+        if (panelPercent != nullptr)
+        {
+            WirePort<Percent> const *portThrottle
+                    = panelPercent->port(MachineRCSController::m_woThrottle);
 
-            float influence = 0.0f;
-            if (commandRot.length() > 0.0f || commandTransl.length() > 0.0f)
+            // Write to throttle output
+            if (portThrottle != nullptr)
             {
-                influence =
-                    thruster_influence(thrusterPos, thrusterDir, commandTransl, commandRot);
+                WireNode<Percent> &nodeThrottle
+                        = rNodesPercent.get_node(portThrottle->m_nodeIndex);
+
+                // Write possibly new throttle value to node
+                SysSignal<Percent>::signal_assign(
+                        rScene, {influence}, nodeThrottle,
+                        portThrottle->m_nodeIndex, updPercent);
             }
-            std::get<wiretype::Percent>(machine.m_woThrottle.value()).m_value = influence;
         }
     }
+
+    // Request to update any wire nodes if they were modified
+    if (!updPercent.empty())
+    {
+        rNodesPercent.update_request(std::move(updPercent));
```

Comment:
Nodes can be marked for update from multiple threads (concurrency not yet implemented though). To reduce the number of times the mutexes are locked, changes are accumulated into a vector (updPercent in this case), and are merged into the main vector after updating.

Reply:
> I still strongly recommend against using mutexes for cross-thread communication.

How much of a problem will this be anyways? A conflict would only happen if two update functions finish at the same time and write to the same nodes which is unlikely.

> Instead, you should break the machines into threads based on some kind of logical partitioning...

This partitioning scheme sounds daunting and complex to implement, especially when noting connections across different wire types. Won't this also limit the number of threads per ship? How often would there even be non-connected parts of the ship? I don't even know where to start with this, nor will it even be any more performant than what I have now.

> If there does need to be some kind of communication between partition, then i recommend doing something like https://codereview.stackexchange.com/questions/219389/c-lock-free-single-producer-single-consumer-queue

Atomic-powered ring buffer? I'll keep this in mind.

> Or alternatively, using a dynamically added task in taskflow that has thread affinity for the thread that the machine in question is running on, or otherwise ensures that the machine in question isn't conncurrently running somehow.

Ordering tasks by using their thread IDs? I'm not sure what you're doing with the machines though.


Anyways, the wiring is already really fast and maximizing its performance isn't really something we need to deal with right now.

Reply:
> How much of a problem will this be anyways? A conflict would only happen if two update functions finish at the same time and write to the same nodes which is unlikely.

I think it will be a noticeable cause of performance issues in the future. Both for the single player game. As well as for multiplayer servers. E.g. multi tenancy related performance problems.

I work on graph processing systems at work all the time and mutexes are frequently a cause of subpar performance. In fact I spent between December and April rewriting a bunch of code to eliminate the use of mutexes entirely because perf testing showed the mutexes being a substantial cause of slowdowns.

My concern is that the wiring system is a **foundational** piece of code for the OSP engine. It needs to be designed very carefully with both single-circuit performance as well as scalability to many many thousands, even many hundreds of thousands of complex circuits. 

Ryzen style "alllllll the threads!" Chips are probably just the beginning of the future explosion of processing cores on a CPU. I very much don't predict that we'll see much growth in terms of single core frequency. So the only practical way to scale up is by scaling out, and mutexes are not going to give good performance for scaling out. Message passing using atomic single produces multi-consumer queues, or explicit job scheduling will scale MUCH better.

And don't think that the single-produces multi-consumer queue is a better solution than using explicit job scheduling (partitioning of the circuits). Its better than mutexes, but it has its own limitations.

Reply:
> This partitioning scheme sounds daunting and complex to implement, especially when noting connections across different wire types. Won't this also limit the number of threads per ship?

You don't want more threads per ship. Making a single ship faster on the wall-clock at the cost of more CPUs doing things us counter productive. You want aggregate throughput. You want to mininize the per frame computation time for when you have 100,000 to 1,000,000 circuits in a scene. The way to do that is to have a circuit calculate to 100% completion in a single thread, with circuits that aren't connected being run in parallel on different threads.

Frankly I don't really think that there should be any cross circuit communications that take less than one frame to complete. So that the circuit can read whatever inputs might come from external to the circuit at the beginning of the frame, and then when the calculations are finished then that circuit doesn't get reactivated until the next frame regardless of events from outside the circuit.

 How often would there even be non-connected parts of the ship? I don't even know where to start with this, nor will it even be any more performant than what I have now.

Reply:
> Anyways, the wiring is already really fast and maximizing its performance isn't really something we need to deal with right now.

Be careful not to build on top of a shakey foundation.

Reply:
Are you suggesting to optimize for numerous small unconnected ships/circuits? I'm expecting our common case to be less than 10 separate ships each with possibly thousands of machines and connections, ie. large crafts with many parts such as space stations. For the case of a large space battle, optimizing for thousands of simpler circuits does make sense.

One thread per ship is reasonably fast, but there will often be cases where one big complex ship will act like a bottleneck. Every other thread of the wire update would just have to wait for this one big ship to finish.

Though I do acknowledge that there are quite a few issues with my systems so far. Multiple threads would end up reading and writing off the same Node buffers. [Signal Nodes here](https://github.com/TheOpenSpaceProgram/osp-magnum/blob/7d5dfb7983fd8bca6959b628bb150abf3fcf63b2/src/osp/Active/SysWire.h#L56
) introduce a false sharing cache miss. `m_read` is read of multiple threads, and `m_write` is written by one thread only while being on the same cache block.

Luckily, this is a good opportunity for me to get rid of the mutexes as well. Instead of just marking nodes for update after writing their values, all changes can just be recorded to a thread-local buffer or atomic queue (or both) that can be fed to the node update afterwards without needing a write member in the node itself.

Partitioning per-vehicle does sound kind of practical; the player would never need to know every circuit element in the world at the same time. The complications I suggest have to do with managing many separate buffers along side EnTT:
* Panel components can be kept unchanged, as they're read-only during wire update.
* Nodes are already stored in vectors, and would probably be easy to separate into each vehicle.
* Machine components may be problematic. They're stored in the registry and need to be modified across multiple threads. However, machines can be made stateless and use only nodes instead (this sounds a little extreme but may really be worth it). 

However, I still believe that a partition system will end up being far more complicated to manage without any or much of a speedup, perhaps even slower if a single ship can bottleneck the entire update.

I've already aggregating the problem, since each type of node and machine are stored in their own buffers. All the data is laid out to be easily processed: every rocket in the scene is one continuous buffer; every fuel tank in the scene is also one continuous buffer. It's simple and easy to manage, and it wouldn't be that hard to have multiple threads process parts of the same buffer if needed. My system doesn't care which vehicles each of these machines belongs to, the only data it's given is the ones that are important.

Reply:
> Are you suggesting to optimize for numerous small unconnected ships/circuits? I'm expecting our common case to be less than 10 separate ships each with possibly thousands of machines and connections, ie. large crafts with many parts such as space stations. For the case of a large space battle, optimizing for thousands of simpler circuits does make sense.

My intention is to say that optimizing for a single ship will come at the cost of aggregate throughput. For example, halving the runtime of one ship, but at the cost of multiple ships having an aggregate runtime increase of 10% is worse over all.

The ship you describe that has many thousands of machines and connections, in my opinion, strongly implies that you are working at the wrong abstraction level. As we talked about, you're thinking like an electrical engineer, but i'm trying to think like a network engineer. Both are valid, but it's my opinion that everything you would want to do can be done with a more abstract model than what you've built, at an improved runtime performance.

Instead of operating at the level of and-gates and or-gates, the use of application specific elements / nodes / items where each such "gate" has application specific code such as multiplexers, routers, ALUs, and so on, provides exactly the same level of "you can make literally anything" power without having the runtime penalty that your approach has, while also:

- Making it much easier to do proper partitioning of disjoint components of the "circuit".
- Not preventing the end user of the game from being silly and plopping down and-gates and similar
- Making it substantially easier for modders and/or players to make interesting circuits without needing to, themselves, be electrical engineers.

For example, an ALU, under the hood, is simply a series of fundamental logic gates (I would know, i had to design one in verilog. ugh.) but no one wants to be designing ALUs in whatever janky visual editor we end up providing. They'd be much much happier plopping down an ALU and calling it good.

> One thread per ship is reasonably fast, but there will often be cases where one big complex ship will act like a bottleneck. Every other thread of the wire update would just have to wait for this one big ship to finish.

That's why we would need to provide to the end user some kind of de-sync item for the circuit to allow them to manually break their ship up into smaller pieces for sake of eliminating bottlenecks like that.

Even though you don't think it's a good idea, i still advocate for an approach that introduces more deliberate latency for gameplay reasons. Though I'm not as concerned about that.

> Luckily, this is a good opportunity for me to get rid of the mutexes as well. Instead of just marking nodes for update after writing their values, all changes can just be recorded to a thread-local buffer or atomic queue (or both) that can be fed to the node update afterwards without needing a write member in the node itself.

I also don't think that needing to mark individual nodes for update is a good strategy either. You'd be better off running the entire circuit based on whether the inputs to that circuit have had any action, and then running the circuit to completion. Doing things at the granularity of individual nodes is going to get expensive quickly, as there will probably end up being millions of nodes in a particularly large multiplayer server.

Reply:
> My intention is to say that optimizing for a single ship will come at the cost of aggregate throughput. For example, halving the runtime of one ship, but at the cost of multiple ships having an aggregate runtime increase of 10% is worse over all.

I don't quite understand why this would be the case. There are N machines to update per frame, scattered across some unspecified number of ships. Ideally, if all the threads are always updating something without waiting, then it would take the same amount of time to finish all updates for that frame regardless of how they're divided between ships. Is there something here I'm missing?

Also note that this is a realtime simulation where all the ships run and update synchronously. They all need to calculate the right values for the physics update, and other systems that need to run afterwards. We can't just update one ship more or less times. There's a fixed minimum number of times its machines need to update.

> The ship you describe that has many thousands of machines and connections, in my opinion, strongly implies that you are working at the wrong abstraction level....

The example I used is a space station, let me elaborate. There would be constant resource flow everywhere: solar panels generating power, oxygen being consumed, fuel being processed etc... There would be a logic system to control some thrusters, docking ports, and maybe some robotic arms. And of course there are spacecraft docked to it, each with maybe a hundred parts. Reaching thousands of machines and connections doesn't sound unrealistic at all.

> Instead of operating at the level of and-gates and or-gates, the use of application specific elements / nodes / items where each such "gate" has application specific code...

This is exactly the level of abstraction I'm already working with. Logic gates are just simple examples. I'm not using booleans here; so far I have Percent and AttitudeControl. Have you looked at the RCSController? Multiplexers, routers, and ALUs is exactly what I have in mind to implement in the future. 

> That's why we would need to provide to the end user some kind of de-sync item for the circuit to allow them to manually break their ship up into smaller pieces for sake of eliminating bottlenecks like that.

When will there ever be a single ship with unconnected sections? I can't think of any practical example where this would happen. Having this de-sync item sounds like an unnecessarily complicated solution for an entirely avoidable problem.

Reply:
> I don't quite understand why this would be the case. There are N machines to update per frame, scattered across some unspecified number of ships. Ideally, if all the threads are always updating something without waiting, then it would take the same amount of time to finish all updates for that frame regardless of how they're divided between ships. Is there something here I'm missing?

The cross-thread synchronization using atomic variables.

If you allow individual machines / nodes in side a single circuit to be updated on multiple threads simultaneously, you must synchronize them to prevent data races.

This means that for a simulation with a single ship, the calculations, spread across multiple threads, will complete more quickly in terms of actual human time, but at the cost of substantially more CPU time across all cores.

When you scale this up to thousands of ships, you're overall throughput for all ships goes down, because of the extra work to synchronize across threads.

> Also note that this is a realtime simulation where all the ships run and update synchronously. They all need to calculate the right values for the physics update, and other systems that need to run afterwards. We can't just update one ship more or less times. There's a fixed minimum number of times its machines need to update.

I don't quite follow where we were discussing updating one ship more or fewer times than another.

But again, i think that you're model is not a good fit if you need to update things multiple times in a single frame. The graph / circuit should be designed in such a way that it will reach a steady state on it's own.

Maybe we need to revisit the the design from a different direction. My proposed system several months ago had each node inside the circuit do a single calculate and a single propagate per frame.

Yours does multiple calculates and propagates per frame

But both of our models had a an external driver that visited the entire collection of nodes in a circuit and poked them one by one.

Maybe we need to instead consider a different approach of using a purely event based circuit, where the circuit receives an input, and that input makes its way through all of the other nodes in the circuit via direct function calls, until all activity in the circuit stops (whether that causes some kind of output or not is irrelevant). Any data storage happens entirely as an internal implementation detail of the nodes in question, and the external driver is ignorant of whether or not nodes store items in EnTT.

This also means that while links between items can be stored in EnTT, data that's passing from one node to another via a link would *not* be stored in EnTT, but instead would involve direct function calls to the next node.

This design allows for cycles (with the explicit requirement that there be some mechanism to prevent recursive function calls blowing out the stack), and it also has the attractive property of the circuit needing to be invoked only when there is an input for it to consume, and the circuit can run as long as it needs to because when it finishes, it's at a steady state until it receives it's next input.

Reply:
> The example I used is a space station, let me elaborate. There would be constant resource flow everywhere: solar panels generating power, oxygen being consumed, fuel being processed etc... There would be a logic system to control some thrusters, docking ports, and maybe some robotic arms. And of course there are spacecraft docked to it, each with maybe a hundred parts. Reaching thousands of machines and connections doesn't sound unrealistic at all.

Sure, that's fair assessment of the number of components / machines that'll be on a particular ship / station. But I'm sceptical that they need to be synchronously connected.

So we have:

- Power generation
- Life support consumption
- Fuel consumption
- Robotics (docking, manipulation, whatever)
- Docked ships interconnected with all the above in some manner.

Power generation, as a "circuit" consumes a "fuel" item (actual fuel, or photons in the case of solar panels) and produces electricity.  That electricity does not need to then synchronously enter the circuit that consumes it. It can be asynchronously posted into some kind of buffer object via a cross-thread message, or a single-producer-single-consumer atomic queue.

The consuming circuit can then asynchronously consume the electricity in it's buffer until it runs out, and when it runs out, it can post an event back to the power producer circuit that it needs more juice.

Similarly, fuel consumption can be handled much the same way. The circuit that requires fuel doesn't need to synchronously pull fuel out of the storage tank. It can have a buffer that's sufficient for it's needs that gets asynchronously filled by the "fuel production" circuit, or "fuel storage" circuit.

Oxygen production and consumption. Water / coolant / sewage and so on, they all have plenty of opportunities for being made asynchronous.

And on a particularly large station, there's no need for all of the electricity circuits to be connected together in one giant rats nest of wires. You'd want a main-bus that has high voltage and current, and then transformers to down-convert the high voltage to a lower voltage for general usage by appliances. Those transformers can, themselves, be async, and represent an opportunity to partition a single station into multiple parallel-y processing circuits.

> This is exactly the level of abstraction I'm already working with. Logic gates are just simple examples. I'm not using booleans here; so far I have Percent and AttitudeControl. Have you looked at the RCSController? Multiplexers, routers, and ALUs is exactly what I have in mind to implement in the future.

I was evaluating it more for C++ code smells than for architectural concerns. My code reviews tend to happen while i'm getting ready for bed on my phone, so it's not always easy to grab a solid architectural understanding.

> When will there ever be a single ship with unconnected sections? I can't think of any practical example where this would happen. Having this de-sync item sounds like an unnecessarily complicated solution for an entirely avoidable problem.

See the proposal above with regards to producers asynchronously inserting into "buffers' (don't have to be buffers in the sense of an std::array, but more in the sense of a "This circuit has this many thingies available right now" counter variable.

Reply:
> When you scale this up to thousands of ships, you're overall throughput for all ships goes down, because of the extra work to synchronize across threads.

How about limiting the number of threads, so there's less things to synchronize? Dividing loads by ship, or dividing by type wouldn't really matter then. I prefer not to run ships individually for other practical reasons I've already described.

> But again, i think that you're model is not a good fit if you need to update things multiple times in a single frame. The graph / circuit should be designed in such a way that it will reach a steady state on it's own.

There are multiple cycles of updates, but each update acts on a different set of machines and nodes each time.

For example, the default part vehicle updates like this when the player yaws, pitches, or rolls the vehicle:
1. Before wire update: MachineUserControl reads inputs for rotation, and requests to modify its AttitudeControl output.
2. Wire update starts
3. Node update: AttitudeControl node writes its new value, and marks connected MachineRCSControllers
4. Calculate Update: MachineRCSControllers update, each reading a new value in AttitudeControl, and requests to modify their connected Throttle outputs
5. Node update: Throttle nodes write new values, and mark the connected MachineRockets to update.
6. Calculate Update: MachineRockets get new throttle values, and calculates their thrust output
7. Wire update ends. No more updates requested. 

> Maybe we need to revisit the the design from a different direction. My proposed system several months ago had each node inside the circuit do a single calculate and a single propagate per frame.

1-frame delays are simply unacceptable for control systems, especially in aerospace. Hovering a drone, landing a booster, or simply controlling by hand would suffer greatly from delays. PID controllers are known to osscilate uncontrollably when there's delay in their feedback. 1-frame for the whole system may be acceptable, but 1 frame for every single machine connected along a line would stack up and become very problematic. This may be acceptable for a different game, but definitly not for level of realism I'm aiming for in this project.

> Maybe we need to instead consider a different approach of using a purely event based circuit, where the circuit receives an input, and that input makes its way through all of the other nodes in the circuit via direct function calls...

This was my original plan for the Urho3D repo. This approach only supports combinational logic, is difficult to manage, unpredictable, and butchers the benefits of having an entity component system by pretty much using it like a heap without the systems. There is also stuff like instruction cache misses everywhere among other issues. Allowing cycles would just be a huge headache. I already have a working system here; if there's a huge glaring flaw with it then it's entirely my fault.

Reply:
> How about limiting the number of threads, so there's less things to synchronize? Dividing loads by ship, or dividing by type wouldn't really matter then. I prefer not to run ships individually for other practical reasons I've already described.

That just means that you reduce framerates because you're not maximizing CPU utilization.


> There is also stuff like instruction cache misses everywhere among other issues. 

I have no idea why you keep indicating that there are instruction cache misses. Where did you measure that?

Basically if you want to submit this, then submit it. I disagree that it's the right approach, but i don't have the time to produce code or examples for a better system beyond what i've already explained.

Reply:
> 1-frame delays are simply unacceptable for control systems, especially in aerospace.

You keep getting hung up on this. Nothing prevented my original proposal from being run multiple times per frame, just like your system does.

Reply:
> ... That electricity does not need to then synchronously enter the circuit that consumes it. It can be asynchronously posted into some kind of buffer object via a cross-thread message, or a single-producer-single-consumer atomic queue

All this just to avoid the 'problem' of being synchronous? Would a buffer really help with multiple machines consuming and producing a resource? From the way you've described it, it really sounds like the buffer object here is a Node.

I strongly believe that there is a really high performance cost in trying to asynchronously manage resource flows like this.

> And lets certainly not forget that unlike electricity, gas and liquid do have noticeable propagation delay properties in pipes / ducts, so treating them as electricity with a "nearly instant" transmission delay isn't an accurate paradigm

Long pipes are best handled with separate machines, not at the system level. There's been a few discussions with z-adams on this topic. Each Node becomes a point of instantaneous pressure when dealing with fluids. A pipe is some machine with some pressure on each end, and an internal state. In electronics, long wires (transmissions lines) are used often in radio frequency design as separate components. Every other connection is just instant.

> And on a particularly large station, there's no need for all of the electricity circuits to be connected together in one giant rats nest of wires. You'd want a main-bus that has high voltage and current, and then transformers to down-convert the high voltage to a lower voltage for general usage by appliances. Those transformers can, themselves, be async, and represent an opportunity to partition a single station into multiple parallel-y processing circuits.

It doesn't matter how cleanly they're connected, they're still connected nonetheless. The idea of having Nodes cleans up a lot of these issues, especially buses. I'm not sure how making these asynchronous would help with anything.

My system handles this issue like this:
1. Calculate update: machines will request to add or subtract resources from a Resource Node by accumulating requests into a thread-local buffer. This buffer is then somehow published to be exposed to the resource node update.
2. Node update: Resource Node will read each request and calculates its net loss or gain. It will then mark its connected producers or consumers to update to replenish or accept resources.

Reply:
> I have no idea why you keep indicating that there are instruction cache misses. Where did you measure that?

I do admit this kind of came out of nowhere. You're implying the use of polymorphism for machines, which doesn't work well with the entity component system. Different types of machines would end up being updated in an unspecified order. A single thread would have to know about many different types of machines dynamically. In my system, one thread only updates one kind of machine. A thread updating rockets will only touch rockets.

> Basically if you want to submit this, then submit it. I disagree that it's the right approach, but i don't have the time to produce code or examples for a better system beyond what i've already explained.

I'm very confident that I've covered a wide enough range of use cases in a flexible and performant manner. It is at least superior to the previous wiring system that this PR replaces. The use of templates here would allow you to entirely implement your mechanics if you want. The 'signals' part of this PR is optional, and I'm working on separating it a bit right now.

Reply:
> All this just to avoid the 'problem' of being synchronous? Would a buffer really help with multiple machines consuming and producing a resource? From the way you've described it, it really sounds like the buffer object here is a Node.
> I strongly believe that there is a really high performance cost in trying to asynchronously manage resource flows like this.

There's two related topics here.

- "multiple circuits being calculated simultaneously" and what the maximum acceptable load the game engine can handle in terms of number of circuits on a particular CPU at e.g. 60 frames per second.
- The runtime cost of a single very large circuit.

Using std::mutex, in either of the situations above, is 100% going to be less performant than some other (non-terribly designed) solution.

For the runtime cost of a single very large circuit, simulating different parts of the circuit on different CPUs sounds attractive because more CPU makes things go quicker. And that's true, up to a point, certainly.

However, that level of cross CPU communication is not going to work well. mutexes cause cache flush events because of the atomic writes, they cause blocking on threads (and potentially sleeping of threads if they can't acquire a lock right away), they cause poor predictability of performance, and framerates that vary maybe not wildly, but certainly more than they would need to.

They're also terrible from a density perspective.

The wall-clock runtime cost of a single very large circuit isn't really relevant when you have many thousands of circuits that all need to be calculated. The aggregate throughput of all circuits in that case is far more important.

The way you get good aggregate throughput for that kind of thing is by having as much "single threaded" calculation happen as you can, by ensuring that the parts of your algorithm / code that touch on other threads are isolated to carefully selected parts of the algorithm. E.g. manually injecting points into the circuits where an asynchrony happens.

> Long pipes are best handled with separate machines, not at the system level. There's been a few discussions with z-adams on this topic. Each Node becomes a point of instantaneous pressure when dealing with fluids. A pipe is some machine with some pressure on each end, and an internal state. In electronics, long wires (transmissions lines) are used often in radio frequency design as separate components. Every other connection is just instant.

That sounds similar to what I was saying, with different terminology.

> It doesn't matter how cleanly they're connected, they're still connected nonetheless. The idea of having Nodes cleans up a lot of these issues, especially buses. I'm not sure how making these asynchronous would help with anything.

It helps in terms of allowing the system to be simulated asynchronously in a predictable manner, to increase aggregate throughput of the overall circuit simulation code.

> My system handles this issue like this: ...........

Which can be a point of asynchronous by having that request / reply happen with a pair of atomic queues.

> I do admit this kind of came out of nowhere. You're implying the use of polymorphism for machines, which doesn't work well with the entity component system. 

I am not implying that currently no. My previous design used polymorphism, which really has no meaningful effect on the instruction cache, but my current concern is with the multi-threaded aspect, which doesn't inherently need polymorphism.

Why do you think polymorphism has any meaningful effect on the instruction cache? That is contrary to my experience in industry and I've never encountered anything to indicate this.

> Different types of machines would end up being updated in an unspecified order. A single thread would have to know about many different types of machines dynamically. In my system, one thread only updates one kind of machine. A thread updating rockets will only touch rockets.

I don't see how this is beneficial at all. What is motivating that as a design goal? Are you saying thread in the meaning of std::thread? If so, how are you handling a situation where there's more types of machines than there are hardware CPU cores? Are you planning to spawn a new std::thread for each of those? Or when taskflow is working for our needs, a new task for each type of machine? That's going to result in absolutely terrible performance.

Using EnTT / ECS is irrelevant, maybe even counter productive, for highly interconnected things like circuits. In an ECS, a particular system should run once, and only once, per frame. That's my opinion anyway, i suppose there could be exceptions to that. But for something like a circuit simulator, or a graph of nodes, ECS is not inherently a useful tool, and you *definitely* should not be doing one "system" per node type if it means needing to run the system over and over again until the nodes reach some steady state.

Reply:
>  It is at least superior to the previous wiring system that this PR replaces.

I complete agree with you on this.

Reply:
> However, that level of cross CPU communication is not going to work well. mutexes cause cache flush events because of the atomic writes, they cause blocking on threads (and potentially sleeping of threads if they can't acquire a lock right away), they cause poor predictability of performance, and framerates that vary maybe not wildly, but certainly more than they would need to.

I'm already in the process of getting rid of mutexes, as well as getting rid of multiple threads writing the same buffer (Nodes) simultaneously. Though the issue with my system is that it constantly needs to pass data between threads after each node or machine update, which happens multiple times per frame. Since each update already touches a different part of memory anyways (noting the part vehicle example), I'm not expecting this to be much of an issue.

> The wall-clock runtime cost of a single very large circuit isn't really relevant when you have many thousands of circuits that all need to be calculated. The aggregate throughput of all circuits in that case is far more important.

Again, numerous small simple circuits won't be our common case. Dividing a large circuit into smaller circuits won't be easy either. 

> Why do you think polymorphism has any meaningful effect on the instruction cache? That is contrary to my experience in industry and I've never encountered anything to indicate this.

I'm repeating things I recall from game developers in the context of ECS and data oriented design. Performance critical sections that rely on virtual functions can suffer from more serious cache misses from needing to load memory from many different types in some unspecified order. It's harder to predict which buffers of memory each function of individual machines read or write into.

> I don't see how this is beneficial at all. What is motivating that as a design goal? Are you saying thread in the meaning of std::thread? If so, how are you handling a situation where there's more types of machines than there are hardware CPU cores? ...

More clearly, each task only updates one type of machine. `MachineRCSController::update_calculate`, for example, will update values for every RCS Controller in the entire scene that needs to be updated. So this is a task for each type. Perhaps even multiple tasks to update the same type in parallel, if enough are marked for update. Where will this terrible performance come from?

> Using EnTT / ECS is irrelevant, maybe even counter productive, for highly interconnected things like circuits. In an ECS, a particular system should run once, and only once, per frame. That's my opinion anyway, i suppose there could be exceptions to that...

It really feels like we're trying to solve entirely different problems here.  I'm using EnTT here to store connections and data, but nothing else I'm doing here is specific to ECS. My scheme doesn't require ECS to work, but EnTT had made many things conveinent. Also, who says an ECS must only run once per frame? This "updating multiple times" thing I keep saying is actually just updaing everything in a certain order (again, noting the part vehicle example where everything is updated only once).

A circuit simulator would benefit from ECS to some extent. Elements in a circuit simulation can be easily represented as entities with components. Stuff like part numbers, temperature, power dissipation, and connections can be components. Implementing higher levels of interconnection is made easier by ECS as it provides a robust method of representing entities. Note that I'm able to tie together logic and resource flows all in one scheme, which would be harder to do without ECS.